app/services/oauth_clients/connection_manager.rb

```diff
@@ -104,28 +103,18 @@ def code_to_token(code)
         update_oauth_client_token(oauth_client_token, service_result.result)
       else
         rack_access_token = service_result.result
-        oauth_client_token =
-          OAuthClientToken.create(
-            user: @user,
-            oauth_client: @oauth_client,
-            origin_user_id: @config.extract_origin_user_id(rack_access_token), # ID of user at OAuth2 Authorization Server
-            access_token: rack_access_token.access_token,
-            token_type: rack_access_token.token_type, # :bearer
-            refresh_token: rack_access_token.refresh_token,
-            expires_in: rack_access_token.raw_attributes[:expires_in],
-            scope: rack_access_token.scope
-          )
-        return ServiceResult.failure(errors: oauth_client_token.errors) unless oauth_client_token.valid?
+        OAuthClientToken.transaction do
+          oauth_client_token = create_client_token(rack_access_token)
 
-        OpenProject::Notifications.send(
-          OpenProject::Events::OAUTH_CLIENT_TOKEN_CREATED,
-          integration_type: @oauth_client.integration_type
-        )
-      end
+          oauth_client_token.save!
```

Comment:
🔴 we need to rollback on a failure here, too. Or is this, what the `!` is for?

Reply:
`save!` raises an error when saving, thus rolling back the transaction.